packages/flutter/lib/src/rendering/editable.dart

```diff
@@ -440,44 +451,100 @@ class RenderEditable extends RenderBox with RelayoutWhenSystemFontsChangeMixin {
 
   void _handleMovement(
       LogicalKeyboardKey key, {
-      @required bool control,
+      @required bool wordModifier,
+      @required bool lineModifier,
       @required bool shift,
     }) {
+    if (wordModifier && lineModifier) {
+      // If both modifiers are down, nothing happens on any of the platforms.
+      return;
+    }
+
     TextSelection newSelection = selection;
 
     final bool rightArrow = key == LogicalKeyboardKey.arrowRight;
     final bool leftArrow = key == LogicalKeyboardKey.arrowLeft;
     final bool upArrow = key == LogicalKeyboardKey.arrowUp;
     final bool downArrow = key == LogicalKeyboardKey.arrowDown;
 
-    // Because the user can use multiple keys to change how they select, the
-    // new offset variable is threaded through these four functions and
-    // potentially changes after each one.
-    if (control) {
-      // If control is pressed, we will decide which way to look for a word
-      // based on which arrow is pressed.
-      if (leftArrow && newSelection.extentOffset > 2) {
-        final TextSelection textSelection = _selectWordAtOffset(TextPosition(offset: newSelection.extentOffset - 2));
-        newSelection = newSelection.copyWith(extentOffset: textSelection.baseOffset + 1);
-      } else if (rightArrow && newSelection.extentOffset < text.toPlainText().length - 2) {
-        final TextSelection textSelection = _selectWordAtOffset(TextPosition(offset: newSelection.extentOffset + 1));
-        newSelection = newSelection.copyWith(extentOffset: textSelection.extentOffset - 1);
+    // Returns true if [c] represents a whitespace code unit.
+    bool isWhitespace(int c) => (c <= 0x000D && c >= 0x0009) || c == 0x0020;
+
+    // Find the previous non-whitespace character
+    int previousNonWhitespace(int extent) {
+      int result = math.max(extent - 1, 0);
+      final String plain = text.toPlainText();
+      while (result > 0 && isWhitespace(plain.codeUnitAt(result))) {
+        result--;
       }
+      return result;
     }
-    // Set the new offset to be +/- 1 depending on which arrow is pressed
-    // If shift is down, we also want to update the previous cursor location
-    if (rightArrow && newSelection.extentOffset < text.toPlainText().length) {
-      newSelection = newSelection.copyWith(extentOffset: newSelection.extentOffset + 1);
-      if (shift) {
-        _cursorResetLocation += 1;
+
+    int nextNonWhitespace(int extent) {
+      final String plain = text.toPlainText();
```

Comment:
Seems like this code would result in use computing plain text every time a key is pressed, which it should be cache-able as long as the TextPainter doesn't update. I'm not sure how large the performance impact would be, but it might be an easy win

Reply:
I added `_plainText` that lazily evaluates and caches the plain text version of the text.  It didn't add too much complexity (actually seems easier to read to me), and can only result in a win performance-wise. The only danger is that the `TextPainter` updates its `text` member somehow without telling the `RenderEditable`, but I don't think that's possible, since the `RenderEditable` owns the painter and doesn't expose it.